pkg/githubauth/app.go

```diff
@@ -0,0 +1,155 @@
+)
+
+// AppScope is an interface which determines the repositories scope for which
+// the app should request tokens.
+type AppScope interface {
+	*allReposScope | *selectedReposScope
+}
+
+type allReposScope struct{}
+
+// ForAllRepos creates a scope that requests permissions on all repositories.
+func ForAllRepos() *allReposScope {
+	return &allReposScope{}
+}
+
+type selectedReposScope struct {
+	repos []string
+}
+
+// ForRepos creates a scope that requests permissions on the given repositories.
+func ForRepos(r ...string) *selectedReposScope {
+	return &selectedReposScope{
+		repos: r,
+	}
+}
+
+var _ TokenSource = (*AppTokenSource[*allReposScope])(nil)
+
+// AppTokenSource is a GitHubToken provider that authenticates as a GitHub App.
+type AppTokenSource[T AppScope] struct {
+	app   *githubapp.GitHubApp
+	scope T
+}
+
+// NewAppTokenSource returns a [AppTokenSource] which authenticates as a GitHub
+// App and returns a GitHub token.
+func NewAppTokenSource[T AppScope](appID, installationID, privateKeyPEM string, scope T) (*AppTokenSource[T], error) {
+	var merr error
+	if appID == "" {
+		merr = errors.Join(merr, fmt.Errorf("missing appID"))
+	}
+	if installationID == "" {
+		merr = errors.Join(merr, fmt.Errorf("missing installationID"))
+	}
+	if privateKeyPEM == "" {
+		merr = errors.Join(merr, fmt.Errorf("missing privateKeyPEM"))
+	}
+	if merr != nil {
+		return nil, fmt.Errorf("failed to create app token source: %w", merr)
+	}
+
+	privateKey, err := parseRSAPrivateKeyPEM(privateKeyPEM)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create app token source: %w", err)
+	}
+
+	app := githubapp.New(githubapp.NewConfig(appID, installationID, privateKey))
+
+	return &AppTokenSource[T]{
+		app:   app,
+		scope: scope,
+	}, nil
+}
+
+func parseRSAPrivateKeyPEM(data string) (*rsa.PrivateKey, error) {
+	block, _ := pem.Decode([]byte(data))
+	if block == nil {
+		return nil, fmt.Errorf("failed to parse pem: no data")
+	}
+
+	key, err := x509.ParsePKCS1PrivateKey(block.Bytes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse private key pem: %w", err)
+	}
+	return key, nil
+}
+
+// GitHubToken implements [TokenSource].
+func (s *AppTokenSource[T]) GitHubToken(ctx context.Context) (string, error) {
+	switch t := any(s.scope).(type) {
+	case *allReposScope:
+		resp, err := s.app.AccessTokenAllRepos(ctx, &githubapp.TokenRequestAllRepos{
+			Permissions: map[string]string{
+				"actions": "read",
+			},
+		})
+		if err != nil {
+			return "", fmt.Errorf("failed to get github access token for all repos: %w", err)
+		}
+		return parseAppTokenResponse(resp)
+
+	case *selectedReposScope:
+		resp, err := s.app.AccessToken(ctx, &githubapp.TokenRequest{
+			Repositories: t.repos,
+			Permissions: map[string]string{
+				"actions": "read",
+			},
+		})
+		if err != nil {
+			return "", fmt.Errorf("failed to get github access token for repos %q: %w", t.repos, err)
+		}
+		return parseAppTokenResponse(resp)
+
+	default:
+		panic("impossible")
```

Comment:
I am curious, is there anything in the pipeline in golang like there is in Java where if all possible cases for a switch are exhausted, `default` doesn't need to be defined? Its one of the features they have been working on for a while now.

Reply:
I haven't been following closely, but I know there are some upcoming improvements to generics.